File: database/database-annotation-processor/src/main/java/ru/tinkoff/kora/database/annotation/processor/RepositoryBuilder.java
package ru.tinkoff.kora.database.annotation.processor;

import com.palantir.javapoet.*;
import jakarta.annotation.Nullable;
import ru.tinkoff.kora.annotation.processor.common.*;
import ru.tinkoff.kora.database.annotation.processor.cassandra.CassandraRepositoryGenerator;
import ru.tinkoff.kora.database.annotation.processor.jdbc.JdbcRepositoryGenerator;

import javax.annotation.processing.ProcessingEnvironment;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.type.DeclaredType;
import java.util.List;

import static ru.tinkoff.kora.annotation.processor.common.TagUtils.makeAnnotationSpec;
import static ru.tinkoff.kora.annotation.processor.common.TagUtils.parseTagValue;

public class RepositoryBuilder {

    private final List<RepositoryGenerator> queryMethodGenerators;

    public RepositoryBuilder(ProcessingEnvironment processingEnv) {
        this.queryMethodGenerators = List.of(
            new JdbcRepositoryGenerator(processingEnv),
            new CassandraRepositoryGenerator(processingEnv)
        );
    }

    @Nullable
    public TypeSpec build(TypeElement repositoryElement) throws ProcessingErrorException {
        var name = NameUtils.generatedType(repositoryElement, "Impl");
        var builder = CommonUtils.extendsKeepAop(repositoryElement, name)
            .addAnnotation(AnnotationUtils.generated(RepositoryAnnotationProcessor.class));

        if (AnnotationUtils.findAnnotation(repositoryElement, CommonClassNames.root) != null) {
            builder.addAnnotation(CommonClassNames.root);
        }
        if (AnnotationUtils.findAnnotation(repositoryElement, CommonClassNames.component) != null) {
            builder.addAnnotation(CommonClassNames.component);
        }

        var tags = parseTagValue(repositoryElement);
        if (!tags.isEmpty()) {
            builder.addAnnotation(makeAnnotationSpec(tags));
        }

        var constructorBuilder = MethodSpec.constructorBuilder()
            .addModifiers(Modifier.PUBLIC);
        if (repositoryElement.getKind().isClass()) {
            this.enrichConstructorFromParentClass(constructorBuilder, repositoryElement);
        }
        var repositoryType = (DeclaredType) repositoryElement.asType();
        for (var availableGenerator : this.queryMethodGenerators) {
            var repositoryInterface = availableGenerator.repositoryInterface();
            var repositoryInterfaceType = TypeUtils.findSupertype(repositoryType, repositoryInterface);

            if (repositoryInterfaceType != null) {
                return availableGenerator.generate(repositoryElement, builder, constructorBuilder);
            }
        }
        throw new ProcessingErrorException("Element doesn't extend any of known repository interfaces", repositoryElement);
    }

    private void enrichConstructorFromParentClass(MethodSpec.Builder constructorBuilder, TypeElement repositoryElement) {
        constructorBuilder.addCode("super(");
        var constructors = CommonUtils.findConstructors(repositoryElement, m -> !m.contains(Modifier.PRIVATE));
        if (constructors.isEmpty()) {
            constructorBuilder.addCode(");\n");
            return;
        }
        if (constructors.size() > 1) {
            throw new ProcessingErrorException("Abstract repository class has more than one public constructor", repositoryElement);
        }
        var constructor = constructors.get(0);
        var parameters = constructor.getParameters();
        for (int i = 0; i < parameters.size(); i++) {
            var parameter = parameters.get(i);
            var constructorParameter = ParameterSpec.builder(TypeName.get(parameter.asType()), parameter.getSimpleName().toString());
            for (var annotation : parameter.getAnnotationMirrors()) {
                constructorParameter.addAnnotation(AnnotationSpec.get(annotation));
            }
            constructorBuilder.addParameter(constructorParameter.build());
            constructorBuilder.addCode("$L", parameter);
            if (i < parameters.size() - 1) {
                constructorBuilder.addCode(", ");
            }
        }
        constructorBuilder.addCode(");\n");
    }

}

File: database/database-symbol-processor/src/main/kotlin/ru/tinkoff/kora/database/symbol/processor/RepositoryBuilder.kt
package ru.tinkoff.kora.database.symbol.processor

import com.google.devtools.ksp.getConstructors
import com.google.devtools.ksp.isProtected
import com.google.devtools.ksp.processing.KSPLogger
import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.symbol.ClassKind
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.squareup.kotlinpoet.AnnotationSpec
import com.squareup.kotlinpoet.CodeBlock
import com.squareup.kotlinpoet.FunSpec
import com.squareup.kotlinpoet.ParameterSpec
import com.squareup.kotlinpoet.TypeSpec
import com.squareup.kotlinpoet.ksp.addOriginatingKSFile
import com.squareup.kotlinpoet.ksp.toAnnotationSpec
import com.squareup.kotlinpoet.ksp.toClassName
import com.squareup.kotlinpoet.ksp.toTypeName
import org.slf4j.LoggerFactory
import ru.tinkoff.kora.database.symbol.processor.cassandra.CassandraRepositoryGenerator
import ru.tinkoff.kora.database.symbol.processor.jdbc.JdbcRepositoryGenerator
import ru.tinkoff.kora.database.symbol.processor.r2dbc.R2DbcRepositoryGenerator
import ru.tinkoff.kora.database.symbol.processor.vertx.VertxRepositoryGenerator
import ru.tinkoff.kora.ksp.common.AnnotationUtils.findAnnotation
import ru.tinkoff.kora.ksp.common.CommonAopUtils.extendsKeepAop
import ru.tinkoff.kora.ksp.common.CommonClassNames
import ru.tinkoff.kora.ksp.common.KspCommonUtils.generated
import ru.tinkoff.kora.ksp.common.exception.ProcessingErrorException
import ru.tinkoff.kora.ksp.common.getOuterClassesAsPrefix

class RepositoryBuilder(
    val resolver: Resolver,
    private val kspLogger: KSPLogger
) {
    private val availableGenerators = listOf(
        JdbcRepositoryGenerator(resolver),
        VertxRepositoryGenerator(resolver, kspLogger),
        R2DbcRepositoryGenerator(resolver),
        CassandraRepositoryGenerator(resolver),
    )
    private val log = LoggerFactory.getLogger(RepositoryBuilder::class.java)


    fun build(repositoryDeclaration: KSClassDeclaration): TypeSpec? {
        log.debug("Generating Repository for {}", repositoryDeclaration.simpleName.asString())
        val name = repositoryDeclaration.getOuterClassesAsPrefix() + repositoryDeclaration.simpleName.asString() + "_Impl"
        val builder = repositoryDeclaration.extendsKeepAop(name, resolver)
            .generated(RepositoryBuilder::class)
            .addOriginatingKSFile(repositoryDeclaration.containingFile!!)

        repositoryDeclaration.findAnnotation(CommonClassNames.tag)
            ?.let { builder.addAnnotation(it.toAnnotationSpec()) }

        val constructorBuilder = FunSpec.constructorBuilder()
        if (repositoryDeclaration.classKind == ClassKind.CLASS) {
            this.enrichConstructorFromParentClass(builder, constructorBuilder, repositoryDeclaration)
        }
        val repositoryType = repositoryDeclaration.asType(listOf())
        for (availableGenerator in this.availableGenerators) {
            val repositoryInterface = availableGenerator.repositoryInterface()
            if (repositoryInterface != null && repositoryInterface.isAssignableFrom(repositoryType)) {
                return availableGenerator.generate(repositoryDeclaration, builder, constructorBuilder)
            }
        }
        throw ProcessingErrorException("Element doesn't extend any of known repository interfaces", repositoryDeclaration)
    }

    private fun enrichConstructorFromParentClass(builder: TypeSpec.Builder, constructorBuilder: FunSpec.Builder, repositoryDeclaration: KSClassDeclaration) {
        val constructors = repositoryDeclaration.getConstructors().filter { !it.isProtected() }.toList()
        if (constructors.isEmpty()) {
            return
        }
        if (constructors.size > 1) {
            throw ProcessingErrorException("Abstract repository class has more than one public constructor", repositoryDeclaration)
        }
        val constructor = constructors[0]
        val parameters = constructor.parameters
        for (i in parameters.indices) {
            val parameter = parameters[i]
            val constructorParameter = ParameterSpec.builder(parameter.name!!.asString(), parameter.type.toTypeName())
            for (annotation in parameter.annotations) {
                val annotationSpec = AnnotationSpec.builder(annotation.annotationType.resolve().toClassName())
                annotation.arguments.forEach { annotationArg ->
                    annotationSpec.addMember(annotationArg.name!!.asString(), annotationArg.value!!)
                }
                constructorParameter.addAnnotation(annotationSpec.build())
            }
            builder.addSuperclassConstructorParameter(CodeBlock.of("%L", parameter.name!!.asString()))
            constructorBuilder.addParameter(constructorParameter.build())
        }
    }
}
